src/strings/ops.c

```diff
@@ -468,16 +471,36 @@ MVMString * MVM_string_concatenate(MVMThreadContext *tc, MVMString *a, MVMString
     if (bgraphs == 0)
         return a;
 
+    is_concat_stable = MVM_nfg_is_concat_stable(tc, a, b);
+    if (is_concat_stable == 0) {
+        MVMCodepoint last_a_first_b[2];
+        last_a_first_b[0] = MVM_string_get_grapheme_at_nocheck(tc, a, a->body.num_graphs - 1);;
+        last_a_first_b[1] = MVM_string_get_grapheme_at_nocheck(tc, b, 0);
+        /* Needing both to be CCC = 0 can probably be relaxed some, but be careful optimizing */
+        if (0 <= last_a_first_b[0] && 0 <= last_a_first_b[1] && MVM_unicode_relative_ccc(tc, last_a_first_b[0]) == 0 && MVM_unicode_relative_ccc(tc, last_a_first_b[1]) == 0) {
+            renormalized_section = MVM_unicode_codepoints_c_array_to_nfg_string(tc, last_a_first_b, 2);
+            MVM_gc_root_temp_push(tc, (MVMCollectable **)&(renormalized_section));
+            consumed_a = 1; consumed_b = 1;
+            renormalized_section_graphs = MVM_string_graphs(tc, renormalized_section);
+            if (agraphs == consumed_a && bgraphs == consumed_b) {
+                MVM_gc_root_temp_pop(tc);
+                return renormalized_section;
+            }
+        }
+    }
     /* Total size of the resulting string can't be bigger than an MVMString is allowed to be. */
     total_graphs = (MVMuint64)agraphs + (MVMuint64)bgraphs;
-    if (total_graphs > MAX_GRAPHEMES)
+    if (total_graphs > MAX_GRAPHEMES) {
+        if (renormalized_section) MVM_gc_root_temp_pop(tc);
         MVM_exception_throw_adhoc(tc,
             "Can't concatenate strings, required number of graphemes %"PRIu64" > max allowed of %lld",
              total_graphs, MAX_GRAPHEMES);
+    }
 
     /* Otherwise, we'll assemble a result string. */
     MVMROOT(tc, a, {
     MVMROOT(tc, b, {
+
```

Comment:
Even if it is NULL in some conditions? renormalized_section is only a string in certain conditions, otherwise it is NULL.